### PR TITLE
Akorade/addsqlfilter

### PR DESCRIPTION
**Pull request by akorade**
Fixes #.

## Changes

Please provide a brief description of the changes here.
1. Added new Filter option to SQL instrumentation for both .NET core and .NET framework.
2. Updated Unit test cases.
3. Updated Readme file

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by utpilla** [comment]
@akorade We need you to sign the CLA before we can accept any contributions. Please follow the instructions for the same.

**Comment by akorade** [comment]
> @akorade We need you to sign the CLA before we can accept any contributions. Please follow the instructions for the same.
done

File: src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientInstrumentationEventSource.cs [hunk 1]
```diff
@@ -75,5 +75,20 @@ public void EnrichmentException(string exception)
         {
             this.WriteEvent(5, exception);
         }
+
+        [Event(6, Message = "Filter threw exception. Request will not be collected. Exception {0}.", Level = EventLevel.Error)]
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientInstrumentationEventSource.cs:79]: this should be `NonEvent`

File: src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientInstrumentationEventSource.cs [hunk 2]
```diff
@@ -75,5 +75,20 @@ public void EnrichmentException(string exception)
...
+            if (this.IsEnabled(EventLevel.Error, EventKeywords.All))
+            {
+                this.RequestFilterException(exception.ToInvariantString());
+            }
+        }
+
+        [Event(6, Message = "Filter threw exception. Request will not be collected. Exception {0}.", Level = EventLevel.Error)]
+        public void RequestFilterException(string exception)
+        {
+            this.WriteEvent(4, exception);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientInstrumentationEventSource.cs:91]:
```suggestion
            this.WriteEvent(6, exception);
```

File: src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientInstrumentationEventSource.cs [hunk 3]
```diff
@@ -75,5 +75,20 @@ public void EnrichmentException(string exception)
...
+        [Event(6, Message = "Filter threw exception. Request will not be collected. Exception {0}.", Level = EventLevel.Error)]
+        public void RequestFilterException(Exception exception)
+        {
+            if (this.IsEnabled(EventLevel.Error, EventKeywords.All))
+            {
+                this.RequestFilterException(exception.ToInvariantString());
+            }
+        }
+
+        [Event(6, Message = "Filter threw exception. Request will not be collected. Exception {0}.", Level = EventLevel.Error)]
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientInstrumentationEventSource.cs:88]:
```suggestion
        [Event(6, Message = "Filter threw exception. Activity will not be collected. Exception {0}.", Level = EventLevel.Error)]
```

File: src/OpenTelemetry.Instrumentation.SqlClient/README.md [hunk 4]
```diff
@@ -60,6 +60,44 @@ For an ASP.NET application, adding instrumentation is typically done in the
...
+
+For .NET Framework Filter receives the payload object containing { (string)DataSource, (string)DatabaseName, (string)CommandText}
+The following code snippet shows how to use Filter with .NET Framework
+
+```csharp
+using var tracerProvider = Sdk.CreateTracerProviderBuilder()
+    .AddSqlClientInstrumentation(opt => opt.Filter
+        = (payload) =>
+    {
+        var dataSource = (string)payload.GetType().GetProperty("DataSource").GetValue(payload, null);
```
Comment (cijothomas) [inline src/OpenTelemetry.Instrumentation.SqlClient/README.md:94]: this usage is quite odd....

Comment (cijothomas) [reply src/OpenTelemetry.Instrumentation.SqlClient/README.md:94]: Could we follow similar model as `Enrich`? (Though that is also a bit odd, but not having reflection is better atleast)

File: src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientInstrumentationEventSource.cs (hunk 1 above)
Comment (akorade) [reply src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlClientInstrumentationEventSource.cs:79]: updated as per comment

File: src/OpenTelemetry.Instrumentation.SqlClient/README.md (hunk 4 above)
Comment (akorade) [reply src/OpenTelemetry.Instrumentation.SqlClient/README.md]: updated as per comment

**Comment by akorade** [comment]
Because of CLA Authorization issue occured due to my Github setup locally with two emails, I am closing this pull request and opening a new one
https://github.com/open-telemetry/opentelemetry-dotnet/pull/3131 with same changes.

**Pull request closed without merging**